analyzers/src/SonarAnalyzer.CSharp/Rules/ClearTextProtocolsAreSensitive.cs

```diff
@@ -0,0 +1,167 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Text.RegularExpressions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+using SonarAnalyzer.SyntaxTrackers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class ClearTextProtocolsAreSensitive : HotspotDiagnosticAnalyzer
+    {
+        private const string DiagnosticId = "S5332";
+        private const string MessageFormat = "Using {0} protocol is insecure. Use {1} instead.";
+        private const string EnableSslMessage = "EnableSsl should be set to true.";
+        private const string HttpPattern = @"^http:\/\/(?!localhost|127.0.0.1).+";
+        private const string FtpPattern = @"^ftp:\/\/.*@(?!localhost|127.0.0.1)";
+        private const string Telnet = "telnet";
+        private const string TelnetPattern = @"^telnet:\/\/.*@(?!localhost|127.0.0.1)";
+        private const string TelnetPatternForIdentifier = @"(Telnet)(?![a-z])";
+        private const string LocalHost = "localhost";
+        private const string LocalHostIp = "127.0.0.1";
+        private const string EnableSslName = "EnableSsl";
+
+        private static readonly DiagnosticDescriptor DefaultRule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager).WithNotConfigurable();
+
+        private static readonly DiagnosticDescriptor EnableSslRule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, EnableSslMessage, RspecStrings.ResourceManager).WithNotConfigurable();
+
+        private readonly Dictionary<string, string> recommendedProtocols = new Dictionary<string, string>
+        {
+            {"telnet", "ssh"},
+            {"ftp", "sftp, scp or ftps"},
+            {"http", "https"},
+            {"clear-text SMTP", "SMTP over SSL/TLS or SMTP with STARTTLS" }
+        };
+
+        private readonly ImmutableArray<string> validServerValues = ImmutableArray.Create(LocalHost, LocalHostIp);
+
+        private readonly CSharpObjectInitializationTracker objectInitializationTracker =
+            new CSharpObjectInitializationTracker(constantValue => constantValue is bool value && value,
+                                                  ImmutableArray.Create(KnownType.System_Net_Mail_SmtpClient, KnownType.System_Net_FtpWebRequest),
+                                                  propertyName => EnableSslName == propertyName);
```

Comment:
This looks like reversed Java convention copy/paste issue.
```suggestion
                                                  propertyName => propertyName == EnableSslName);
```